src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs

```diff
@@ -1113,8 +1114,10 @@ private static int HandleEventPeerRecvAborted(State state, ref QUIC_STREAM_EVENT
                     shouldShutdownWriteComplete = true;
                 }
 
-                state.SendState = SendState.Aborted;
                 state.SendErrorCode = (long)streamEvent.PEER_RECEIVE_ABORTED.ErrorCode;
+                // make sure the SendErrorCode above is commited to memory before we assign the state. This
+                // ensures that the code is read correctly in SetupWriteStartState when checking without lock
+                Volatile.Write(ref Unsafe.As<SendState, int>(ref state.SendState), (int)SendState.Aborted);
```

Comment:
I don't think you need the `Volatile.Write` here since you're inside a lock. AFAIK that should emit memory barriers if necessary. @stephentoub please correct me if I'm wrong.

Reply:
The comment suggests the concern is the reordering of the write to SendErrorCode and the write to SendState. The volatile is necessary to prevent that reordering. There is a fence as part of releasing the lock, but that only guarantees that neither write will move past the lock exit, not that they won't move past each other.

Reply:
:facepalm: yep, I should have read that comment :rofl: